lisa/tests/staging/utilclamp.py

```diff
@@ -0,0 +1,311 @@
+                key=lambda band: band[1] - band[0],
+                reverse=True
+            )
+            bands = bands[:nr_bands]
+            bands = sorted(bands, key=itemgetter(0))
+
+        return bands
+
+    @classmethod
+    def _get_phases(cls, plat_info):
+        """
+        Returns a list of phases. Each phase being described by a tuple:
+
+          (uclamp_val, util, (band))
+        """
+        capacities = cls._collect_capacities_flatten(plat_info)
+        bands = cls._get_bands(capacities)
+
+        def band_mid(band):
+            return int((band[1] + band[0]) / 2)
+
+        return [
+                (
+                    band_mid(band),
+                    band_mid(band) / 2,
+                    band
+                )
+                for band in bands
+        ]
+
+    @classmethod
+    def get_rtapp_profile(cls, plat_info):
+        periods = [
+            Periodic(
+                duty_cycle_pct=(util / PELT_SCALE) * 100,  # util to pct
+                duration_s=5,
+                period_ms=cls.TASK_PERIOD_MS,
+                uclamp_min=uclamp_val,
+                uclamp_max=uclamp_val
+            )
+            for uclamp_val, util, expected in cls._get_phases(plat_info)
+        ]
+
+        return {'task': functools.reduce(lambda a, b: a + b, periods)}
+
+    def _get_task_activation_mean_capacities(self, task, start, end):
+        df = self.trace.analysis.tasks.df_task_activation(task)
+        df = df[['active', 'cpu']]
+
+        df_freq = self.trace.analysis.frequency.df_cpus_frequency()
+        df_freq = df_freq[['cpu', 'frequency']]
+        df_freq = df_freq.pivot(index=None,
+                                columns='cpu',
+                                values='frequency').reset_index()
+        df_freq.set_index('Time', inplace=True)
+        df = df.merge(df_freq, how='outer', left_index=True, right_index=True)
+
+        end = end if not np.isnan(end) else df.last_valid_index()
+
+        # Skip the first half to skip any task migration
+        start += (end - start) / 2
+
+        # Find the first activation
+        start = df[(df.index >= start) &
+                   (df['active'] == 1)].first_valid_index()
+
+        if (start >= end):
+            raise ValueError('Phase ends before it has even started')
+
+        df = df.ffill()
+        df = df_window(df, (start, end), method='exclusive')
+
+        # Fake Time, representing only the task activation period
+        df_add_delta(df, col='Time', inplace=True)
+        df = df.dropna()
+        df = df[df.active == 1]
+        df['Time'] = df['Time'].cumsum()
+
+        # Frequency to capacity
+        capacities = self._collect_capacities(self.plat_info)
+        for cpu, freq_to_capa in capacities.items():
+            df[cpu] = df[cpu].map(freq_to_capa)
+
+        df['capacity'] = df.apply(lambda line: line[line.cpu], axis=1)
+        df['capacity_max'] = df.apply(
+                                lambda line:
```

Comment:
too much indentation. Once the unneeded indentation is removed, the lambda fits on its own line:
```
        df['capacity_max'] = df.apply(
            lambda line: max(capacities[line.cpu].values()),
            axis=1
        )
```

Reply:
That code is gone